src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceProvider.cs

```diff
@@ -19,50 +18,16 @@ public sealed class ServiceProvider : IServiceProvider, IDisposable, IServicePro
 
         private readonly CallSiteValidator _callSiteValidator;
 
-        internal ServiceProvider(IEnumerable<ServiceDescriptor> serviceDescriptors, ServiceProviderOptions options)
+        internal ServiceProvider(IEnumerable<ServiceDescriptor> serviceDescriptors, IServiceProviderEngine engine, ServiceProviderOptions options)
         {
-            IServiceProviderEngineCallback callback = null;
+            _engine = engine;
+
             if (options.ValidateScopes)
             {
-                callback = this;
+                _engine.InitializeCallback(this);
                 _callSiteValidator = new CallSiteValidator();
             }
 
-            switch (options.Mode)
-            {
-                case ServiceProviderMode.Default:
-#if !NETCOREAPP
-                    _engine = new DynamicServiceProviderEngine(serviceDescriptors, callback);
-#else
-                    if (RuntimeFeature.IsSupported("IsDynamicCodeCompiled"))
```

Comment:
This wasn't a test feature. We fallback to non-compiled implementations on platforms that don't support runtime code compilation.